src/3dxdisp.c

```diff
@@ -49,6 +49,24 @@ int tdx_open(void)
 	return 0;
 }
 
+int tdx_open_file(const char *dev_path)
+{
+	tdx_close();
+
+	if (dev_path == NULL) {
+		dev_path = "/dev/spacenav";
+	}
+
+	if(!(hid = hid_open_path(dev_path))) {
+		fprintf(stderr, "tdx_open: failed to open space pilot HID device\n");
+		return -1;
+	}
+	hid_set_nonblocking(hid, 1);
+
+	return 0;
+}
+
+
```

Comment:
I agree about adding a tdx_open_file function for more flexibility, but I don't see how this is making anything simpler. The tdx_open function just looks it up by the vendor:device ids, which can't ever fail if the device is connected, and needs no further arguments. Is there some corner case I'm missing, in which this would be simpler?

Also I would change the null check to call `tdx_open()` instead, again because it can't fail, while the link might not exist if the user did not use our suggested udev rule.

Reply:
If can i like to skip enumerating devices. 

without udev rule, program must run as root, its not good solution.

with udev rule, we dont need to enumerate, just use symlink.

Reply:
I'm sure finding a device by ID is an instantaneous operation. The udev rule to avoid running as root is certainly ideal, but the simplest use case is the most general use case. One which works with or without the udev rule, and with or without a symlink. I like the open_file for flexibility, but it's clearly not the simplest, nor the most foolproof method.